_drafts/partially-ordering-time.md

```diff
@@ -0,0 +1,253 @@
+their team lays out an analysis of the C++ weak memory model in a slight specialisation of their framework.
+
+This gives us a convenient way to discuss the similarities and differences between the two
+fields. In particular, the C++ memory model _approximates_ to per-object causal
+consistency, with a few interesting differences.
+
+We'll study them axiom by axiom, but it's first important to note that where causal
+consistency is defined in (Burckhardt's) terms of a “visibility” and an “arbitration”
+orderings, they define weak memory axioms in terms of a “reads-from” and “modification
+order” instead, correspondingly.
+
+“Visibility” in the distributed systems model is a partial order of potential causality;
+it tracks what events (on possibly other replicas) are visible to what other events. It
+can contain read→read, read→write, write→read, and write→write edges, across multiple
+objects. “Reads-from” is much more specialised; it only ever contains write→read edges on
+the same object, and only ever contains one edge for a given read event.
+
+This corresponds to the fact that there only is a single actual memory cell being written
+to in a shared memory processor, for any given object, even if threads may access it at
+different points in causality. In a distributed system, a logical object can be written to
+at many separate locations (“replicas”).
+
+“Arbitration” is a total order that tracks how the distributed system, when asked to make
+a choice, will adjudicate which event happened before which other event. It's distinct
+from “modification order” in that the latter is, again, per-object, and only adjudicates
+writes.
+
+This is, again, a specialisation borne from the fact that the weak memory specification
+only makes strong guarantees per-object.
+
+Okay. With these under our belt, we can discuss each individual property and how it
+differs between distributed systems' per-object causal consistency and the C++11 weak
+memory model.
+
+#### EVENTUAL
+
+For any particular event, there cannot be infinitely many events that do not have
+visibility of it. To make sense under weak memory, this is slightly more complicated: you
+have to state that for any particular _write_, there cannot be infinitely many reads that
+don't read from it or read from an earlier write (by modification order).
+
+The C++11 specification does not guarantee this axiom, though practically it's difficult
+to find a counterexample.
+
+#### THINAIR
+
+When you trace “potential causality” across thread/client operations and
+visibility/reads-from, you cannot go back in time. This is just a sanity axiom in
+distributed systems, but it does prohibit certain kinds of speculative execution from
+being visible to the user in weak memory systems.
+
+The C++11 specification does not guarantee this axiom, and it is violated on commodity
+hardware consistently.
+
+#### Causality Axioms
```

Comment:
I think everything from here down reads pretty well. I'll re-visit after I've looked at the rendered version